### PR TITLE
Command upgrade

### PR DESCRIPTION
**Pull request by KarolDenst**

### CODE REVIEW

File: GraphAlgorithms/Commands/CliqueCommand.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+public static class CliqueCommand
+{
+    public static Command Create()
+    {
+        var fileOption = OptionsFactory.CreateFileOption();
+        var indexOption = OptionsFactory.CreateIndexOption();
+        var algorithmTypeOption = OptionsFactory.CreateAlgorithmTypeOption();
+        var cmpOption = OptionsFactory.CreateCmpOption();
+
+        var command = new Command("max-clique", "Find maximum clique is a graph.")
```
Comment (AleksyBalazinski) [inline GraphAlgorithms/Commands/CliqueCommand.cs]: typo: `is` -> `in`

File: GraphAlgorithms/Commands/CliqueCommand.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
+﻿using System.CommandLine;
+using GraphAlgorithms.Clique;
+using GraphAlgorithms.Graphs;
```
Comment (AleksyBalazinski) [inline GraphAlgorithms/Commands/CliqueCommand.cs]: unused

File: GraphAlgorithms/Commands/CliqueCommand.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        {
+            "vertices" => finder.Find(graph),
+            "vertices-then-edges" => finder.FindWithEdges(graph),
+            _ => throw new NotImplementedException()
+        };
+
+        Console.WriteLine(string.Join(", ", clique.ToArray()));
+        Console.WriteLine($"Number of vertices: {clique.Count}");
+        Console.WriteLine();
+        Console.WriteLine(graph.ToString(clique));
```
Comment (AleksyBalazinski) [inline GraphAlgorithms/Commands/CliqueCommand.cs]: I don't know if we should print it like that. If I recall correctly, prof was talking about something like adjacency matrix for the **whole** input graph with clique marked by a different color, so for 
```
1
5
0 0 1 0 1
1 0 0 0 0
1 0 0 0 1
0 0 0 0 0
1 0 1 0 0
```
we would have
```
1
5
0 0 *1 0 *1
1 0 0 0 0
*1 0 0 0 *1
0 0 0 0 0
*1 0 *1 0 0
```
(replace `*` with color)

File: GraphAlgorithms/Commands/GraphCommand.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
+﻿using System.CommandLine;
+
+namespace GraphAlgorithms.Commands;
+
+public static class GraphCommand
+{
+    public static Command Create()
```
Comment (AleksyBalazinski) [inline GraphAlgorithms/Commands/GraphCommand.cs:7]: unused?

File: GraphAlgorithms/Commands/MetricCommand.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+{
+    public static Command Create()
+    {
+        var fileOption = OptionsFactory.CreateFileOption();
+        var file2Option = OptionsFactory.CreateFile2Option();
+        var index1Option = OptionsFactory.CreateIndex1Option();
+        var index2Option = OptionsFactory.CreateIndex2Option();
+        var algorithmTypeOption = OptionsFactory.CreateAlgorithmTypeOption();
+
+        var command = new Command("metric", "Finds the distance between two graphs.")
```
Comment (AleksyBalazinski) [inline GraphAlgorithms/Commands/MetricCommand.cs]: Let's be consistent and change it to infinitive ie `Find the distance ...`

File: GraphAlgorithms/Commands/MetricCommand.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
+﻿using System.CommandLine;
+using GraphAlgorithms.Clique;
+using GraphAlgorithms.MCS;
+using GraphAlgorithms.Utils;
```
Comment (AleksyBalazinski) [inline GraphAlgorithms/Commands/MetricCommand.cs]: unused

File: GraphAlgorithms/Commands/MetricCommand.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        var graph1 = Utils.GetGraph(path1, index1);
+        var graph2 = Utils.GetGraph(path2, index2);
+
+        ICliqueFastFinder finder = GetFinder(algType);
+        bool withEdges = false;
+
+        (int[] subgraph1, int[] subgraph2) = MCSFinder.FindFast(graph1, graph2, finder, withEdges);
+
+        var metric = 1 - (double)subgraph1.Length / Math.Max(graph1.Size, graph2.Size);
```
Comment (AleksyBalazinski) [inline GraphAlgorithms/Commands/MetricCommand.cs:48]: duplicate from `Utils.Metrics.cs`

File: GraphAlgorithms/Commands/OptionsFactory.cs [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        index1Option.AddAlias("-i1");
+
+        return index1Option;
+    }
+
+    public static Option<int> CreateIndex2Option()
+    {
+        var index2Option = new Option<int>(name: "--index2",
+                description: "Index of the 2nd graph in the file",
+                getDefaultValue: () => 1);
```
Comment (AleksyBalazinski) [inline GraphAlgorithms/Commands/OptionsFactory.cs]: maybe change the default to 0

File: GraphAlgorithms/Commands/SizeCommand.cs [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+        command.SetHandler(Run, fileOption, indexOption);
+
+        return command;
+    }
+
+    private static void Run(string path, int index)
+    {
+        var graph = Utils.GetGraph(path, index);
+
+        Console.WriteLine($"Graph size: ({graph.Size}, {graph.GetNumberOfEdges()})");
```
Comment (AleksyBalazinski) [inline GraphAlgorithms/Commands/SizeCommand.cs:26]: this is not regarding the code, but just a side note: we probably should write in the document that we define size like that and we order them "lexicographically" (or whatever the clever word for that was)

File: GraphAlgorithms/Commands/SubgraphCommand.cs [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        var graph2 = Utils.GetGraph(path2, index2);
+
+        ICliqueFastFinder finder = GetFinder(algType);
+        bool withEdges = comparison == "vertices-then-edges";
+
+        (int[] subgraph1, int[] subgraph2) = MCSFinder.FindFast(graph1, graph2, finder, withEdges);
+
+        Console.WriteLine("MCS in graph 1: " + string.Join(", ", subgraph1));
+        Console.WriteLine("MCS in graph 2: " + string.Join(", ", subgraph2));
+        Console.WriteLine($"{subgraph1.Length} vertices");
```
Comment (AleksyBalazinski) [inline GraphAlgorithms/Commands/SubgraphCommand.cs:52]: I think we should print the adjacency matrices for small graphs with subgraphs marked (similarly for clique)

File: GraphAlgorithms/Program.cs [hunk 11]
```diff
@@ -17,271 +12,23 @@ private static void Main(string[] args)
...
-                (subgraph1, subgraph2) = MCSFinder.FindFast(graphs[index1], graphs[index2], new MaxCliqueHeuFinder(), withEdges: true);
-            }
-            else throw new NotImplementedException();
-        }
-        else throw new NotImplementedException();
-
-        Console.WriteLine("MCS in graph 1: " + string.Join(", ", subgraph1));
-        Console.WriteLine("MCS in graph 2: " + string.Join(", ", subgraph2));
-        Console.WriteLine($"{subgraph1.Length} vertices");
+        return rootCommand;
```
Comment (AleksyBalazinski) [inline GraphAlgorithms/Program.cs:54]: making a stupid mistake like giving a path to non-existing file causes very nasty error message about unhandled exception:
image
I think we should catch all exceptions and just print error message from the exception

File: GraphAlgorithms/Commands/MetricCommand.cs (hunk 7 above)
Comment (KarolDenst) [reply GraphAlgorithms/Commands/MetricCommand.cs:48]: not quite. Here I give an option to use a heuristic finder

**Review by AleksyBalazinski — APPROVED** [review]

**Pull request merged**